README.md
# octal_clock

[![Codemagic build status](https://api.codemagic.io/apps/63336f980c4aeb133f6a4413/63336f980c4aeb133f6a4412/status_badge.svg)](https://codemagic.io/apps/63336f980c4aeb133f6a4413/63336f980c4aeb133f6a4412/latest_build)
[![Pub Version](https://img.shields.io/pub/v/octal_clock)](https://pub.dev/packages/octal_clock)

A dart library to manage converting from standard times to octal time.

Octal time does away with the cumbersome 12-based imperial system of time keeping, where you can't count using only your fingers and everything is weird and awkward.
By switching to octal time, we get all the benefits of a metric-style system, and keep the benefit of being able to easily half and quarter time measurements.

The basics of the system are as follows: \
1000<sub>8</sub> microseconds per millisecond. \
1000<sub>8</sub> milliseconds per second. \
100<sub>8</sub> seconds per minute. \
100<sub>8</sub> octal minutes per hour. \
10<sub>8</sub> octal hours per scisma. \
3 scismas per day.

A "scisma" is much like AM/PM, except there are three of them: Morning (Mane Scisma/MS), Day (Dies Scisma/DS), Evening (Vesperum Scisma/VS).
For the purposes of simplification Mane Scisma starts at 12:00 AM imperial time.

Why 3 scismas? \
In the slightly modified words of Robert Owen, "\[1 Scisma]'s labour, \[1 Scisma]'s recreation, \[1 Scisma]'s rest". \
... Also, because 30<sub>8</sub> is 24<sub>10</sub>, so all the hours line up and our math is pretty and we don't need to change the calendar.

## Usage

A simple usage example:
```dart
import 'package:octal_clock/octal_clock.dart';

main() {
  var octalDate = OctalDateTime.now();
  print('Octal Time: $octalDate');

  // Even though we're passing in an int, we treat it as if it were octal
  var octalDuration = OctalDuration(minutes: 74);
  var future = octalDate.add(octalDuration);
  print('The future: $future');

  var octalUtc = octalDate.toUtc();
  var utcAsLocal = OctalDateTime(
      octalUtc.year,
      octalUtc.month,
      octalUtc.day,
      octalUtc.hour,
      octalUtc.minute,
      octalUtc.second,
      octalUtc.millisecond,
      octalUtc.microsecond);

  // Run this multiple time to observe the conversion errors
  var offset = utcAsLocal.difference(octalDate);
  print('Octal UTC offset: $offset');
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/flutter-institute/lib_octal_clock.dart/issues


CHANGELOG.md
# Changelog

## 3.0.0

- Updating Dart SDK to 3.0.0
- Removing unused anlyzer options

## 2.0.0

- Null safety
- Fixing a bunch of linter errors
- Removing unneeded `new` and `const` keywords in tests
- Properly adding `@override` annotations in tests

## 1.0.0+1

- Pub scoring updates
- Adding dartdoc comments to every public member
- Pedantic static analysis fixes
- Converting all double quoted strings to single quotes
- Removing unneeded `this` for accessing members
- Removing unneeded `new` and `const` keywords
- Properly adding `@override` annotations
- Fixing default value assignment for OctalDuration from `:` to `=`
- Removing unneeded type annotations for local variables

## 1.0.0

- Upgrading from dart2 compatibility
- Renaming all the OctalDuration constants to match the Duration class (breaking change)

## 0.0.5

- Exporting the converter.dart util so it can be used by others

## 0.0.4

- Fixing typo in `HOURS_PER_SCISMA` constant

## 0.0.3

- Adding a constants for `SCISMAS_PER_DAY` and `HOURS_PER_SCISMA`

## 0.0.2

- Updating my name in the license comments

## 0.0.1

- Initial version
- Basic DateTime and Duration implementations


test/octal_clock_test.dart
// Copyright (c) 2017, Brian Armstrong. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

import 'package:octal_clock/octal_clock.dart';
import 'package:test/test.dart';

void _expectDate(odt, int year, int month, int day, int hour, int minute,
    int second, int millisecond, int microsecond, bool isUtc) {
  expect(odt.year, year, reason: 'year did not match');
  expect(odt.month, month, reason: 'month did not match');
  expect(odt.day, day, reason: 'day did not match');
  expect(odt.hour, hour, reason: 'hour did not match');
  expect(odt.minute, minute, reason: 'minute did not match');
  expect(odt.second, second, reason: 'second did not match');
  expect(odt.millisecond, millisecond, reason: 'millisecond did not match');
  expect(odt.microsecond, microsecond, reason: 'microsecond did not match');
  expect(odt.isUtc, isUtc, reason: 'isUtc did not match');
}

void _expectDuration(od, int days, int hours, int minutes, int seconds,
    int milliseconds, int microseconds) {
  expect(od.inDays, days, reason: 'days did not match');
  expect(od.inHours, hours, reason: 'hours did not match');
  expect(od.inMinutes, minutes, reason: 'minutes did not match');
  expect(od.inSeconds, seconds, reason: 'seconds did not match');
  expect(od.inMilliseconds, milliseconds, reason: 'milliseconds did not match');
  expect(od.inMicroseconds, microseconds, reason: 'microseconds did not match');
}

void main() {
  group('OctalDateTime', () {
    group('::new', () {
      test('gets midnight correct', () {
        _expectDate(
            OctalDateTime(2017, 10, 31), 2017, 10, 31, 0, 0, 0, 0, 0, false);
      });

      test('gets midnigth correct utc', () {
        _expectDate(
            OctalDateTime.utc(2017, 10, 31), 2017, 10, 31, 0, 0, 0, 0, 0, true);
      });

      test('handles milliseconds correctly', () {
        _expectDate(OctalDateTime(2017, 1, 1, 0, 0, 0, 777, 777), 2017, 1, 1, 0,
            0, 0, 777, 777, false);
      });

      test('gets MS correctly', () {
        _expectDate(OctalDateTime.utc(2017, 10, 31, 7), 2017, 10, 31, 7, 0, 0,
            0, 0, true);
        _expectDate(OctalDateTime(2017, 10, 31, 10), 2017, 10, 31, 10, 0, 0, 0,
            0, false);
        _expectDate(OctalDateTime.utc(2017, 10, 31, 11), 2017, 10, 31, 11, 0, 0,
            0, 0, true);
      });

      test('gets DS correctly', () {
        _expectDate(OctalDateTime(2017, 10, 31, 17, 77, 77, 777, 777), 2017, 10,
            31, 17, 77, 77, 777, 777, false);
        _expectDate(OctalDateTime.utc(2017, 10, 31, 20), 2017, 10, 31, 20, 0, 0,
            0, 0, true);
        _expectDate(OctalDateTime(2017, 10, 31, 21), 2017, 10, 31, 21, 0, 0, 0,
            0, false);
      });

      test('gets VS correctly', () {
        _expectDate(OctalDateTime.utc(2017, 10, 31, 27), 2017, 10, 31, 27, 0, 0,
            0, 0, true);
        _expectDate(
            OctalDateTime(2017, 10, 31, 30), 2017, 11, 1, 0, 0, 0, 0, 0, false);
      });

      test('errors on invalid octal numbers', () {
        expect(
          () => OctalDateTime(2017, 10, 31, 29),
          throwsA(allOf(
              isArgumentError,
              predicate((dynamic e) =>
                  e.message == 'hour (29) is not a valid octal number.'))),
        );
        expect(
          () => OctalDateTime(2017, 10, 31, 27, 18),
          throwsA(allOf(
              isArgumentError,
              predicate((dynamic e) =>
                  e.message == 'minute (18) is not a valid octal number.'))),
        );
        expect(
          () => OctalDateTime(2017, 10, 31, 27, 15, 68),
          throwsA(allOf(
              isArgumentError,
              predicate((dynamic e) =>
                  e.message == 'second (68) is not a valid octal number.'))),
        );
        expect(
          () => OctalDateTime(2017, 10, 31, 27, 15, 45, 182),
          throwsA(allOf(
              isArgumentError,
              predicate((dynamic e) =>
                  e.message ==
                  'millisecond (182) is not a valid octal number.'))),
        );
        expect(
          () => OctalDateTime(2017, 10, 31, 27, 15, 45, 74, 935),
          throwsA(allOf(
              isArgumentError,
              predicate((dynamic e) =>
                  e.message ==
                  'microsecond (935) is not a valid octal number.'))),
        );
      });
    });

    group('::fromDateTime', () {
      test('converts correctly', () {
        final dt = DateTime.utc(2017, 10, 31, 8, 50, 37, 450, 268);
        _expectDate(OctalDateTime.fromDateTime(dt), 2017, 10, 31, 10, 65, 77,
            743, 211, true);
      });
    });

    // TODO ::toDateTime

    group('::parse', () {
      test('parses octal date correctly', () {
        _expectDate(OctalDateTime.parse('2017-10-31 10:65:77.743211Z'), 2017,
            10, 31, 10, 65, 77, 743, 211, true);
        _expectDate(OctalDateTime.parse('2017-12-31 27:77:77.777777'), 2017, 12,
            31, 27, 77, 77, 777, 777, false);
      });

      test('errors on invalid octal', () {
        expect(
          () => OctalDateTime.parse('2017-10-31 18:35:21'),
          throwsA(allOf(
            isArgumentError,
            predicate((dynamic e) =>
                e.message == 'hour (18) is not a valid octal number.'),
          )),
        );
        expect(
          () => OctalDateTime.parse('2017-10-31 17:38:21'),
          throwsA(allOf(
            isArgumentError,
            predicate((dynamic e) =>
                e.message == 'minute (38) is not a valid octal number.'),
          )),
        );
        expect(
          () => OctalDateTime.parse('2017-10-31 17:35:29'),
          throwsA(allOf(
            isArgumentError,
            predicate((dynamic e) =>
                e.message == 'second (29) is not a valid octal number.'),
          )),
        );
        expect(
          () => OctalDateTime.parse('2017-10-31 17:35:27.182123'),
          throwsA(allOf(
            isArgumentError,
            predicate((dynamic e) =>
                e.message ==
                'milli/microsecond (182123) is not a valid octal number.'),
          )),
        );
        expect(
          () => OctalDateTime.parse('2017-10-31 17:35:27.123128'),
          throwsA(allOf(
            isArgumentError,
            predicate((dynamic e) =>
                e.message ==
                'milli/microsecond (123128) is not a valid octal number.'),
          )),
        );
      });
    });

    test('microseconds since epoch', () {
      var odt = OctalDateTime.utc(1970, 1, 1, 0, 0, 0);
      expect(odt.millisecondsSinceEpoch, 0);
      expect(odt.microsecondsSinceEpoch, 0);

      odt = OctalDateTime.utc(1970, 1, 1, 0, 0, 0, 1);
      expect(odt.millisecondsSinceEpoch, 1);
      expect(odt.microsecondsSinceEpoch, 1000);

      odt = OctalDateTime.utc(1970, 1, 1, 0, 0, 0, 500, 500);
      expect(odt.millisecondsSinceEpoch, 500);
      expect(odt.microsecondsSinceEpoch, 500500);

      odt = OctalDateTime.utc(1970, 1, 1, 0, 0, 0, 0, 1);
      expect(odt.millisecondsSinceEpoch, 0);
      expect(odt.microsecondsSinceEpoch, 1);
    });

    test('::toString', () {
      var odt = OctalDateTime(2017, 10, 31, 26, 73, 65, 123, 456);
      expect(odt.toString(), '2017-10-31 26:73:65.123456');
      odt = OctalDateTime.utc(2017, 10, 31, 26, 73, 65, 123, 456);
      expect(odt.toString(), '2017-10-31 26:73:65.123456Z');
    });

    test('::toIso8601String', () {
      var odt = OctalDateTime(2017, 10, 31, 26, 73, 65, 123, 456);
      expect(odt.toIso8601String(), '2017-10-31T26:73:65.123456');
      odt = OctalDateTime.utc(2017, 10, 31, 26, 73, 65, 123, 456);
      expect(odt.toIso8601String(), '2017-10-31T26:73:65.123456Z');
    });
  });

  group('OctalDuration', () {
    test('creates a full duration', () {
      var duration = OctalDuration(
          days: 1,
          hours: 2,
          minutes: 3,
          seconds: 4,
          milliseconds: 5,
          microseconds: 6);
      _expectDuration(duration, 1, 32, 3203, 320304, 320304005, 320304005006);
    });

    test('handles multiple days', () {
      var duration = OctalDuration(days: 2, seconds: 1);
      _expectDuration(duration, 2, 60, 6000, 600001, 600001000, 600001000000);
    });

    test('handles adding durations', () {
      var first = OctalDuration(days: 1);
      _expectDuration(first, 1, 30, 3000, 300000, 300000000, 300000000000);
      var second = OctalDuration(
          hours: 2, minutes: 3, seconds: 4, milliseconds: 5, microseconds: 6);
      _expectDuration(second, 0, 2, 203, 20304, 20304005, 20304005006);
      _expectDuration(
          first + second, 1, 32, 3203, 320304, 320304005, 320304005006);
    });

    test('handles subtracting durations', () {
      var first = OctalDuration(
          days: 1,
          hours: 2,
          minutes: 3,
          seconds: 4,
          milliseconds: 5,
          microseconds: 6);
      var second = OctalDuration(days: 1, seconds: 4);
      _expectDuration(first - second, 0, 2, 203, 20300, 20300005, 20300005006);
      _expectDuration(
          second - first, 0, -2, -203, -20300, -20300005, -20300005006);
    });

    test('handles multiplying durations', () {
      var duration = OctalDuration(days: 1, seconds: 1);
      _expectDuration(
          duration * 2, 2, 60, 6000, 600002, 600002000, 600002000000);

      duration = OctalDuration(
          days: 1,
          hours: 2,
          minutes: 3,
          seconds: 4,
          milliseconds: 5,
          microseconds: 6);
      _expectDuration(
          duration * 3, 3, 116, 11611, 1161114, 1161114017, 1161114017022);
    });

    test('handles dividing durations', () {
      var duration = OctalDuration(days: 2, seconds: 2, microseconds: 3);
      _expectDuration(
          duration ~/ 2, 1, 30, 3000, 300001, 300001000, 300001000001);
    });

    test('handle comparisons', () {
      var oneday = OctalDuration(days: 1);
      var twoday = OctalDuration(days: 2);
      var thirtyhours = OctalDuration(hours: 30);
      var sixtyhours = OctalDuration(hours: 60);

      // Less Than
      expect(oneday < twoday, true);
      expect(twoday < oneday, false);
      expect(oneday < oneday, false);
      expect(oneday < thirtyhours, false);

      // Greater Than
      expect(twoday > oneday, true);
      expect(oneday > twoday, false);
      expect(twoday > twoday, false);
      expect(twoday > sixtyhours, false);

      // Less Than Equal
      expect(oneday <= twoday, true);
      expect(twoday <= oneday, false);
      expect(oneday <= oneday, true);
      expect(oneday <= thirtyhours, true);

      // Greater Than Equal
      expect(twoday >= oneday, true);
      expect(oneday >= twoday, false);
      expect(twoday >= twoday, true);
      expect(twoday >= sixtyhours, true);
    });

    test('::toString', () {
      var duration = OctalDuration(
          days: 2,
          hours: 3,
          minutes: 14,
          seconds: 5,
          milliseconds: 6,
          microseconds: 777);
      expect(duration.toString(), '63:14:05.006777');

      duration = OctalDuration(
          days: -2,
          hours: -3,
          minutes: -14,
          seconds: -5,
          milliseconds: -6,
          microseconds: -777);
      expect(duration.toString(), '-63:14:05.006777');
    });

    test('::abs', () {
      var duration = OctalDuration(
          days: -2,
          hours: -3,
          minutes: -14,
          seconds: -5,
          milliseconds: -6,
          microseconds: -777);
      expect(duration.abs().toString(), '63:14:05.006777');
    });

    test('::fromDuration', () {
      var duration = OctalDuration.fromDuration(Duration(days: 1, hours: 1));
      _expectDuration(duration, 1, 31, 3100, 310000, 310000000, 310000000000);

      duration =
          OctalDuration.fromDuration(Duration(days: 1, hours: 1, minutes: 1));
      _expectDuration(duration, 1, 31, 3101, 310104, 310104211, 310104211000);

      duration = OctalDuration.fromDuration(
          Duration(days: 1, hours: 1, minutes: 1, seconds: 1));
      _expectDuration(duration, 1, 31, 3101, 310105, 310105317, 310105317000);

      duration = OctalDuration.fromDuration(
          Duration(days: 1, hours: 1, minutes: 1, seconds: 1, milliseconds: 1));
      _expectDuration(duration, 1, 31, 3101, 310105, 310105320, 310105320000);

      duration = OctalDuration.fromDuration(Duration(
          days: 1,
          hours: 8,
          minutes: 10,
          seconds: 59,
          milliseconds: 999,
          microseconds: 998));
      _expectDuration(duration, 1, 40, 4013, 401356, 401356735, 401356735777);
    });

    test('::toDuration', () {
      var duration = OctalDuration(days: 1, hours: 1);
      _expectDuration(
          duration.toDuration(), 1, 25, 1500, 90000, 90000000, 90000000000);

      duration = OctalDuration(days: 1, hours: 1, minutes: 1);
      _expectDuration(
          duration.toDuration(), 1, 25, 1500, 90056, 90056250, 90056250000);

      duration = OctalDuration(days: 1, hours: 1, minutes: 1, seconds: 1);
      _expectDuration(
          duration.toDuration(), 1, 25, 1500, 90057, 90057129, 90057129000);

      duration = OctalDuration(
          days: 1, hours: 1, minutes: 1, seconds: 1, milliseconds: 1);
      _expectDuration(
          duration.toDuration(), 1, 25, 1500, 90057, 90057131, 90057131000);

      duration = OctalDuration(
          days: 1,
          hours: 1,
          minutes: 1,
          seconds: 1,
          milliseconds: 1,
          microseconds: 1);
      _expectDuration(
          duration.toDuration(), 1, 25, 1500, 90057, 90057131, 90057131001);
    });
  });

  group('OctalDateTime manipulation with OctalDuration', () {
    test('adds durations to OctalDateTimes', () {
      var odt = OctalDateTime(2017, 10, 31, 25, 63, 74, 123, 456);
      _expectDate(odt, 2017, 10, 31, 25, 63, 74, 123, 456, false);
      odt = odt.add(OctalDuration(days: 1));
      _expectDate(odt, 2017, 11, 1, 25, 63, 74, 123, 456, false);
      odt = odt.add(OctalDuration(hours: 2));
      _expectDate(odt, 2017, 11, 1, 27, 63, 74, 123, 456, false);
      odt = odt.add(OctalDuration(minutes: 3));
      _expectDate(odt, 2017, 11, 1, 27, 66, 74, 123, 456, false);
      odt = odt.add(OctalDuration(seconds: 4));
      _expectDate(odt, 2017, 11, 1, 27, 67, 0, 123, 456, false);
      odt = odt.add(OctalDuration(milliseconds: 5));
      // If we were doing direct manipulation, millis would be 130
      // In this case, we are running in to the `round`ing between imperial and octal
      _expectDate(odt, 2017, 11, 1, 27, 67, 0, 131, 456, false);
      odt = odt.add(OctalDuration(microseconds: 6));
      // We lose a lot of accuracy going back and forth with imperial microseconds
      // Micros should be 464, but our loss of precision in rounding hurts us here
      // We could minimize the rounding error by going for 010000 microseconds, but that would be weird
      _expectDate(odt, 2017, 11, 1, 27, 67, 0, 131, 460, false);
    });
  });
}


lib/src/octal_date_time.dart
// Copyright (c) 2017, Brian Armstrong. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

// A thinly veiled proxy around dart.core's DateTime that is
// Copyright (c) 2011, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

part of octal_clock;

/// Convert an octal value into a decimal value
/// [octal] the octal value to convert
/// [type] a string describing this octal value, used for context in error messages
/// [max] the maximum decimal value allowed
int toDecimal(int octal, String type, [int? max]) {
  try {
    final decimal = oct2dec(octal);
    if (max != null && decimal > max) {
      throw ArgumentError(
          '$type ($octal) cannot be larger than ${dec2oct(max)}.');
    }
    return decimal;
  } on InvalidOctalNumber {
    throw ArgumentError('$type ($octal) is not a valid octal number.');
  }
}

DateTime _octalPartsToDateTime(int year, int month, int day, int hour,
    int minute, int second, int millisecond, int microsecond,
    [bool isUtc = false]) {
  final octalMillis =
      toDecimal(hour, 'hour') * OctalDuration.millisecondsPerHour +
          toDecimal(minute, 'minute') * OctalDuration.millisecondsPerMinute +
          toDecimal(second, 'second') * OctalDuration.millisecondsPerSecond +
          toDecimal(millisecond, 'millisecond');

  var imperialMillis = (octalMillis / MILLISECONDS_RATIO).round();
  final imperialHour = imperialMillis ~/ Duration.millisecondsPerHour;
  imperialMillis =
      imperialMillis.remainder(Duration.millisecondsPerHour).toInt();
  final imperialMinute = imperialMillis ~/ Duration.millisecondsPerMinute;
  imperialMillis =
      imperialMillis.remainder(Duration.millisecondsPerMinute).toInt();
  final imperialSecond = imperialMillis ~/ Duration.millisecondsPerSecond;
  imperialMillis =
      imperialMillis.remainder(Duration.millisecondsPerSecond).toInt();
  final imperialMicroseconds = (toDecimal(microsecond, 'microsecond',
              OctalDuration.microsecondsPerMillisecond) /
          MICROSECONDS_RATIO)
      .round();

  DateTime imperialEquivalent;
  if (isUtc) {
    imperialEquivalent = DateTime.utc(year, month, day, imperialHour,
        imperialMinute, imperialSecond, imperialMillis, imperialMicroseconds);
  } else {
    imperialEquivalent = DateTime(year, month, day, imperialHour,
        imperialMinute, imperialSecond, imperialMillis, imperialMicroseconds);
  }

  return imperialEquivalent;
}

/// A class representing an DateTime in Octal format
class OctalDateTime implements Comparable<OctalDateTime> {
  final DateTime _date;
  final int _millisFromEpoch;
  final int _millis;

  late int _hour;
  late int _minute;
  late int _second;
  late int _millisecond;
  late int _microsecond;

  /// Create a octal datetime from octal parts in the user's timezone
  OctalDateTime(int year,
      [int month = 1,
      int day = 1,
      int hour = 0,
      int minute = 0,
      int second = 0,
      int millisecond = 0,
      int microsecond = 0])
      : this.fromDateTime(_octalPartsToDateTime(year, month, day, hour, minute,
            second, millisecond, microsecond, false));

  /// Create a octal datetime from octal parts in UTC
  OctalDateTime.utc(int year,
      [int month = 1,
      int day = 1,
      int hour = 0,
      int minute = 0,
      int second = 0,
      int millisecond = 0,
      int microsecond = 0])
      : this.fromDateTime(_octalPartsToDateTime(year, month, day, hour, minute,
            second, millisecond, microsecond, true));

  /// Create a octal datetime representing the current moment
  OctalDateTime.now() : this.fromDateTime(DateTime.now());

  /// Convert an imperial [DateTime] into an octal datetime
  OctalDateTime.fromDateTime(DateTime dt)
      : _date = dt,
        _millisFromEpoch =
            (dt.millisecondsSinceEpoch * MILLISECONDS_RATIO).round(),
        _millis = ((dt.hour * Duration.millisecondsPerHour +
                    dt.minute * Duration.millisecondsPerMinute +
                    dt.second * Duration.millisecondsPerSecond +
                    dt.millisecond) *
                MILLISECONDS_RATIO)
            .round() {
    final parts = millisecondsToParts(_millis);
    _hour = parts[0];
    _minute = parts[1];
    _second = parts[2];
    _millisecond = parts[3];

    _microsecond = (_date.microsecond * MICROSECONDS_RATIO).round();
  }

  /// Convert octal [milliseconds] into their larger parts
  /// returns a list(hours, minutes, seconds, milliseconds) representing the total miliseconds value
  static List<int> millisecondsToParts(int milliseconds) {
    final hours = milliseconds ~/ OctalDuration.millisecondsPerHour;
    milliseconds =
        milliseconds.remainder(OctalDuration.millisecondsPerHour).toInt();
    final minutes = milliseconds ~/ OctalDuration.millisecondsPerMinute;
    milliseconds =
        milliseconds.remainder(OctalDuration.millisecondsPerMinute).toInt();
    final seconds = milliseconds ~/ OctalDuration.millisecondsPerSecond;
    milliseconds =
        milliseconds.remainder(OctalDuration.millisecondsPerSecond).toInt();

    return [hours, minutes, seconds, milliseconds];
  }

  /// Convert this octal datetime into an imperial [DateTime]
  DateTime toDateTime() {
    return DateTime.fromMillisecondsSinceEpoch(
        (_millisFromEpoch / MILLISECONDS_RATIO).round());
  }

  /// Compare this octal datetime to [other]
  /// Returns -1 if this is before other
  /// Returns 1 if this is after other
  /// Returns 0 if they are the same
  @override
  int compareTo(OctalDateTime other) {
    return _date.compareTo(other._date);
  }

  /// Check if this octal datetime is equal to [other]
  @override
  bool operator ==(other) {
    if (!(other is OctalDateTime)) return false;
    return _date == other._date;
  }

  /// Get the hascode representation of this octal datetime
  @override
  int get hashCode => _date.hashCode;

  /// Check if this octal datetime is after [other]
  bool isAfter(OctalDateTime other) {
    return _date.isAfter(other._date);
  }

  /// Check if this octal datetime if the same time as [other]
  bool isAtSameMomentAs(OctalDateTime other) {
    return _date.isAtSameMomentAs(other._date);
  }

  /// Check if this octal datetime is before [other]
  bool isBefore(OctalDateTime other) {
    return _date.isBefore(other._date);
  }

  /// Convert this octal datetime into the local timezone
  OctalDateTime toLocal() {
    return OctalDateTime.fromDateTime(_date.toLocal());
  }

  /// Convert this octal datetime into UTC
  OctalDateTime toUtc() {
    return OctalDateTime.fromDateTime(_date.toUtc());
  }

  /// Add [duration] to this octal datetime
  OctalDateTime add(OctalDuration duration) {
    return OctalDateTime.fromDateTime(_date.add(duration.toDuration()));
  }

  /// Subtract [duration] from this octal datetime
  OctalDateTime subtract(OctalDuration duration) {
    return OctalDateTime.fromDateTime(_date.subtract(duration.toDuration()));
  }

  /// Get the difference between this octal datetime and [other]
  /// Returns an [OctalDuration] representing the difference
  OctalDuration difference(OctalDateTime other) {
    return OctalDuration.fromDuration(_date.difference(other._date));
  }

  /// Returns true if this octal datetime is represented in UTC, false otherwise
  bool get isUtc => _date.isUtc;

  /// Get the octal hour portion of the octal datetime
  int get hour => dec2oct(_hour);

  /// Get the octal minute portion of the octal datetime
  int get minute => dec2oct(_minute);

  /// Get the octal second portion of the octal datetime
  int get second => dec2oct(_second);

  /// Get the octal millisecond portion of the octal datetime
  int get millisecond => dec2oct(_millisecond);

  /// Get the octal milliseconds that have elapsed since the epoch
  int get millisecondsSinceEpoch => dec2oct(_millisFromEpoch);

  /// Get the octal microsecond  portion of the octal datetime
  int get microsecond => dec2oct(_microsecond);

  /// Get the octal microseconds that have elapsed since the epoch
  int get microsecondsSinceEpoch =>
      dec2oct(_millisFromEpoch * OctalDuration.microsecondsPerMillisecond +
          _microsecond);

  /// Get the month portion of the octal datetime
  int get month => _date.month;

  /// Get the day portion of the octal datetime
  int get day => _date.day;

  /// Get the year portion of the octal datetime
  int get year => _date.year;

  /// Get the name of the timezone for the octal datetime
  String get timeZoneName => _date.timeZoneName;

  /// Get the timezone offset of this octal duration
  /// Returns an [OctalDuration] representing this value
  OctalDuration get timeZoneOffset =>
      OctalDuration.fromDuration(_date.timeZoneOffset);

  /// Get the weekday for this octal datetime
  int get weekday => _date.weekday;

  static String _fourDigits(int n) {
    final absN = n.abs();
    final sign = n < 0 ? '-' : '';
    if (absN >= 1000) return '$n';
    if (absN >= 100) return '${sign}0$absN';
    if (absN >= 10) return '${sign}00$absN';
    return '${sign}000$absN';
  }

  static String _sixDigits(int n) {
    assert(n < -9999 || n > 9999);
    final absN = n.abs();
    final sign = n < 0 ? '-' : '+';
    if (absN >= 100000) return '$sign$absN';
    return '${sign}0$absN';
  }

  static String _threeDigits(int n) {
    if (n >= 100) return '$n';
    if (n >= 10) return '0$n';
    return '00$n';
  }

  static String _twoDigits(int n) {
    if (n >= 10) return '$n';
    return '0$n';
  }

  /// Get the simple string representation of this octal datetime
  @override
  String toString() {
    final y = _fourDigits(year);
    final m = _twoDigits(month);
    final d = _twoDigits(day);
    final h = _twoDigits(hour);
    final min = _twoDigits(minute);
    final sec = _twoDigits(second);
    final ms = _threeDigits(millisecond);
    final us = microsecond == 0 ? '' : _threeDigits(microsecond);
    if (isUtc) {
      return '$y-$m-$d $h:$min:$sec.$ms${us}Z';
    } else {
      return '$y-$m-$d $h:$min:$sec.$ms$us';
    }
  }

  /// Get the ISO8601 string representation of this octal datetime
  String toIso8601String() {
    final y =
        (year >= -9999 && year <= 9999) ? _fourDigits(year) : _sixDigits(year);
    final m = _twoDigits(month);
    final d = _twoDigits(day);
    final h = _twoDigits(hour);
    final min = _twoDigits(minute);
    final sec = _twoDigits(second);
    final ms = _threeDigits(millisecond);
    final us = microsecond == 0 ? '' : _threeDigits(microsecond);
    if (isUtc) {
      return '$y-$m-${d}T$h:$min:$sec.$ms${us}Z';
    } else {
      return '$y-$m-${d}T$h:$min:$sec.$ms$us';
    }
  }

  /// Parse an octal datimetime from an ISO8601 string
  /// [formattedString] the ISO8601 representation of the octal datetime
  static OctalDateTime parse(String formattedString) {
    /*
     * date ::= yeardate time_opt timezone_opt
     * yeardate ::= year colon_opt month colon_opt day
     * year ::= sign_opt digit{4,6}
     * colon_opt :: <empty> | ':'
     * sign ::= '+' | '-'
     * sign_opt ::=  <empty> | sign
     * month ::= digit{2}
     * day ::= digit{2}
     * time_opt ::= <empty> | (' ' | 'T') hour minutes_opt
     * minutes_opt ::= <empty> | colon_opt digit{2} seconds_opt
     * seconds_opt ::= <empty> | colon_opt digit{2} millis_opt
     * micros_opt ::= <empty> | '.' digit{1,6}
     * timezone_opt ::= <empty> | space_opt timezone
     * space_opt :: ' ' | <empty>
     * timezone ::= 'z' | 'Z' | sign digit{2} timezonemins_opt
     * timezonemins_opt ::= <empty> | colon_opt digit{2}
     */
    final re = RegExp(r'^([+-]?\d{4,6})-?(\d\d)-?(\d\d)' // Day part.
        r'(?:[ T](\d\d)(?::?(\d\d)(?::?(\d\d)(?:\.(\d{1,6}))?)?)?' // Time part.
        r'( ?[zZ]| ?([-+])(\d\d)(?::?(\d\d))?)?)?$'); // Timezone part.

    final match = re.firstMatch(formattedString);
    if (match != null) {
      int parseIntOrZero(String? matched) {
        if (matched == null) return 0;
        return int.parse(matched);
      }

      int validOctal(int value, {int? max, String? type}) {
        try {
          final decimal = oct2dec(value);
          if (max != null && decimal > max) {
            throw ArgumentError(
                '$type ($value) cannot be larger than ${dec2oct(max)}.');
          }
          return value;
        } on InvalidOctalNumber {
          throw ArgumentError('$type ($value) is not a valid octal number.');
        }
      }

      // Parses fractional second digits of '.(\d{1,6})' into the combined
      // microseconds.
      int parseMilliAndMicroseconds(String? matched) {
        if (matched == null) return 0;
        final length = matched.length;
        assert(length >= 1);
        assert(length <= 6);

        var result = 0;
        for (var i = 0; i < 6; i++) {
          result *= 10;
          if (i < matched.length) {
            result += matched.codeUnitAt(i) ^ 0x30;
          }
        }
        return validOctal(result,
            type: 'milli/microsecond',
            max: OctalDuration.microsecondsPerSecond);
      }

      final year = int.parse(match[1]!);
      final month = int.parse(match[2]!);
      final day = int.parse(match[3]!);
      final hour = validOctal(parseIntOrZero(match[4]),
          type: 'hour', max: OctalDuration.hoursPerDay);
      var minute = validOctal(parseIntOrZero(match[5]),
          type: 'minute', max: OctalDuration.minutesPerHour);
      final second = validOctal(parseIntOrZero(match[6]),
          type: 'second', max: OctalDuration.secondsPerMinute);
      final milliAndMicroseconds = oct2dec(parseMilliAndMicroseconds(match[7]));
      final millisecond = dec2oct(
          milliAndMicroseconds ~/ OctalDuration.microsecondsPerMillisecond);
      final microsecond = dec2oct(milliAndMicroseconds
          .remainder(OctalDuration.microsecondsPerMillisecond)
          .toInt());
      var isUtc = false;
      if (match[8] != null) {
        // timezone part
        isUtc = true;
        if (match[9] != null) {
          // timezone other than 'Z' and 'z'.
          final sign = (match[9] == '-') ? -1 : 1;
          final hourDifference = int.parse(match[10]!);
          var minuteDifference = parseIntOrZero(match[11]);
          minuteDifference += 60 * hourDifference;
          minute -= sign * minuteDifference;
        }
      }

      final imperialEquivalent = _octalPartsToDateTime(year, month, day, hour,
          minute, second, millisecond, microsecond, isUtc);

      return OctalDateTime.fromDateTime(imperialEquivalent);
    } else {
      throw FormatException('Invalid date format', formattedString);
    }
  }
}
